subprojects/plugins/src/integTest/groovy/org/gradle/jvm/component/internal/DefaultJvmSoftwareComponentIntegrationTest.groovy

```diff
@@ -198,13 +226,312 @@ class DefaultJvmSoftwareComponentIntegrationTest extends AbstractIntegrationSpec
+
+            task verify {
+                assert components.comp
+                assert components.comp.features.collect { it.name } == ['feat']
+                assert components.comp.features.feat.variants.collect { it.name } == ['featApiElements', 'featJavadocElements', 'featRuntimeElements', 'featSourcesElements']
+                assert components.comp.variants.collect { it.name } == ['featApiElements', 'featJavadocElements', 'featRuntimeElements', 'featSourcesElements']
+            }
+        """
+
+        expect:
+        succeeds "verify"
+    }
+
+    def "component exposes variants from multiple features"() {
+        given:
+        buildFile << """
+            plugins {
+                id('java-base')
+            }
+
+            ${factoryRegistrationGroovy()}
+
+            components {
+                comp(JvmSoftwareComponent) {
+                    features {
+                        feat1(SingleTargetJvmFeature) {
+                            withJavadocJar()
+                            withSourcesJar()
+                        }
+                        feat2(SingleTargetJvmFeature)
+                    }
+                }
+            }
+
+            task verify {
+                assert components.comp
+                assert components.comp.features.collect { it.name } == ['feat1', 'feat2']
+                assert components.comp.features.feat1.variants.collect { it.name } == ['feat1ApiElements', 'feat1JavadocElements', 'feat1RuntimeElements', 'feat1SourcesElements']
+                assert components.comp.features.feat2.variants.collect { it.name } == ['feat2ApiElements', 'feat2RuntimeElements']
+                assert components.comp.variants.collect { it.name } == ['feat1ApiElements', 'feat1JavadocElements', 'feat1RuntimeElements', 'feat1SourcesElements', 'feat2ApiElements', 'feat2RuntimeElements']
+            }
+        """
+
+        expect:
+        succeeds "verify"
+    }
+
+    def "can configure custom variants for single target features"() {
+        mavenRepo.module("org", "foo", "1.0").publish()
+
+        given:
+        buildFile << """
+            plugins {
+                id('java-base')
+            }
+
+            ${factoryRegistrationGroovy()}
+            repositories { maven { url '${mavenRepo.uri}' } }
+
+            configurations {
+                customConfiguration
+            }
+
+            dependencies {
+                customConfiguration 'org:foo:1.0'
+            }
+
+            components {
+                comp(JvmSoftwareComponent) {
+                    features {
+                        feat(SingleTargetJvmFeature) {
+                            variants {
+                                customRuntimeElements(ConfigurationBackedConsumableVariant) {
+                                    configuration {
+                                        extendsFrom configurations.customConfiguration
+                                    }
+                                }
```

Comment:
💭 If extending is going to the be usual idiom for creating these, maybe we can further sub-type to make the DSL even simpler and provide something like:

```groovy
variants {
    customRuntimeElements(ExtendingConfigurationBackedConsumableVariant, configurations.customConfiguration)
}
```

Reply:
This was just one simple example. I think users using these configuration-backed variants will often be extending, declaring attributes, _and_ adding artifacts.